File: ringstrings-app/src/test/java/com/coreman2200/ringstrings/symbol/inputprocessor/numerology/NumerologicalChartProcessor.java
package com.coreman2200.ringstrings.symbol.inputprocessor.numerology;

import com.coreman2200.ringstrings.profiledata.IProfileDataBundle;
import com.coreman2200.ringstrings.protos.RingStringsAppSettings;
import com.coreman2200.ringstrings.symbol.chart.NumerologicalChartImpl;
import com.coreman2200.ringstrings.symbol.inputprocessor.numerology.grouped.GroupedNumberSymbolsInputProcessor;
import com.coreman2200.ringstrings.symbol.inputprocessor.numerology.grouped.IGroupedNumberSymbolsInputProcessor;
import com.coreman2200.ringstrings.symbol.numbersymbol.grouped.GroupedNumberSymbols;
import com.coreman2200.ringstrings.symbol.numbersymbol.interfaces.IChartedNumberSymbols;
import com.coreman2200.ringstrings.symbol.numbersymbol.interfaces.IGroupedNumberSymbols;

/**
 * NumerologicalChartProcessor
 * Builds a full Numerological Chart for a Profile.
 *
 * Created by Cory Higginbottom on 5/28/15
 * http://github.com/coreman2200
 *
 * Licensed under the GNU General Public License (GPL), Version 2.0.
 * You may not use this file except in compliance with this License.
 *
 * You may obtain a copy of the GPLv2 License at
 * http://www.gnu.org/licenses/old-licenses/gpl-2.0.html
 */

public class NumerologicalChartProcessor extends GroupedNumberSymbolsInputProcessor implements IGroupedNumberSymbolsInputProcessor {

    public NumerologicalChartProcessor(RingStringsAppSettings settings) {
        super(settings);
    }

    public IChartedNumberSymbols produceGroupedNumberSymbolsForProfile(IProfileDataBundle profile) {
        assert (profile != null);
        userProfile = profile;
        IChartedNumberSymbols chart = new NumerologicalChartImpl();

        // Grouped Symbols that establish the full number symbol
        final GroupedNumberSymbols[] groupedSymbols = { GroupedNumberSymbols.QUALITIES, GroupedNumberSymbols.CHALLENGES,
                                                        GroupedNumberSymbols.PERIODS, GroupedNumberSymbols.PERSONAL,
                                                        GroupedNumberSymbols.PINNACLES };

        for (GroupedNumberSymbols type : groupedSymbols)
            chart.addNumberSymbol(type, getGroupedNumberSymbolsForType(type));

        return chart;
    }

    private IGroupedNumberSymbols getGroupedNumberSymbolsForType(GroupedNumberSymbols type) {
        return genProcessorType(type).produceGroupedNumberSymbolsForProfile(userProfile);
    }
    
    protected IGroupedNumberSymbolsInputProcessor genProcessorType(GroupedNumberSymbols type) {
        return GroupedNumberSymbolsInputProcessor.getProcessor(mAppSettings, type);
    }

}

File: domain/src/main/java/com/coreman2200/ringstrings/domain/input/numerology/NumerologicalChartProcessor.kt
package com.coreman2200.ringstrings.domain.input.numerology

import com.coreman2200.ringstrings.domain.NumerologySettings
import com.coreman2200.ringstrings.domain.input.numerology.grouped.GroupedNumberSymbolsInputProcessor
import com.coreman2200.ringstrings.domain.input.numerology.grouped.IGroupedNumberSymbolsInputProcessor
import com.coreman2200.ringstrings.domain.input.entity.IProfileData
import com.coreman2200.ringstrings.domain.symbol.numbersymbol.grouped.GroupedNumbers
import com.coreman2200.ringstrings.domain.symbol.numbersymbol.impl.GroupedNumberSymbol
import com.coreman2200.ringstrings.domain.symbol.numbersymbol.impl.NumerologicalChart

/**
 * NumerologicalChartProcessor
 * Builds a full Numerological Chart for a Profile.
 *
 * Created by Cory Higginbottom on 5/28/15
 * http://github.com/coreman2200
 *
 * Licensed under the GNU General Public License (GPL), Version 2.0.
 * You may not use this file except in compliance with this License.
 *
 * You may obtain a copy of the GPLv2 License at
 * http://www.gnu.org/licenses/old-licenses/gpl-2.0.html
 */
class NumerologicalChartProcessor(
    profile: IProfileData,
    settings: NumerologySettings
) : GroupedNumberSymbolsInputProcessor(profile, settings), IGroupedNumberSymbolsInputProcessor {
    override fun produceGroupedNumberSymbolsForProfile(): NumerologicalChart {
        val chart = NumerologicalChart()
        chart.profileid = profile.id

        // Grouped Symbols that establish the full number symbol
        val groupedSymbols: Array<GroupedNumbers> = arrayOf<GroupedNumbers>(
            GroupedNumbers.QUALITIES, GroupedNumbers.CHALLENGES,
            GroupedNumbers.PERIODS, GroupedNumbers.PERSONAL,
            GroupedNumbers.PINNACLES
        )
        for (type in groupedSymbols) {
            val grouped = getGroupedNumberSymbolsForType(type)
            grouped.getAll().forEach {
                it.chartid = chart.chartid
                it.profileid = chart.profileid
            }

            chart.add(type, grouped)
        }
        return chart
    }

    private fun getGroupedNumberSymbolsForType(type: GroupedNumbers): GroupedNumberSymbol {
        return genProcessorType(type).produceGroupedNumberSymbolsForProfile()
    }

    private fun genProcessorType(type: GroupedNumbers): IGroupedNumberSymbolsInputProcessor {
        return getProcessor(type, profile, settings)
    }
}
